### PR TITLE
Minimizing need to delete provisioned volume

### PR DESCRIPTION
**Pull request by sbezverk**
Current behaviour of PV controller, if by some reason PV object cannot be saved by API server is to delete just provisioned volume. It is not desired behaviour because provision/de-provision operations could be long for some storage backends. This PR changes this behaviour and proposes to store successfully provisioned volume information as a annotation in PVC object.  On PV object save failure, the controller will re-attempt to bind PVC to PV, but it will not need to provision a volume since it has already been done. It needs just to retrieve PV definition from PVC's annotation. On any error, PV controller will switch to old logic.

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

```release-note
NONE
```

### CODE REVIEW

**Comment by jsafrane** [comment]
There are several issues in this PR:

1. PV contains sensitive information, like ID of the volume, that can't be edited by regular users that can create PVC. With this pr, rogue user could craft a PVC with `annStoredVolumeData` with ID of some else's volume and steal data.

2. If saving PV fails due to API server throttling, it's very likely that patching PVC will fail too.

If we have troubles with in-tree volumes (and I haven't noticed any), I would propose to add exponential backoff to writing PV, now it tries 5x with 10 sec sleep  in between.

**Comment by sbezverk** [comment]
@jsafrane Would it be acceptable solution if I find a way to to encrypt `annStoredVolumeData`?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ import (
 	"strings"
 	"time"
```
Comment (wgliang) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: There should be no empty line.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 2]
```diff
@@ -1451,36 +1462,43 @@ func (ctrl *PersistentVolumeController) provisionClaimOperation(claim *v1.Persis
...
-		// so just let controller retry in the next sync. We'll only call func
-		// rescheduleProvisioning here when the underlying provisioning actually failed.
-		ctrl.rescheduleProvisioning(claim)
-
-		strerr := fmt.Sprintf("Failed to provision volume with StorageClass %q: %v", storageClass.Name, err)
-		glog.V(2).Infof("failed to provision volume for claim %q with StorageClass %q: %v", claimToClaimKey(claim), storageClass.Name, err)
-		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.ProvisioningFailed, strerr)
-		return
+	volRecovered := false
+	if claim.ObjectMeta.Annotations[annVolumeAlreadyProvisioned] == "yes" {
```
Comment (wgliang) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: if annotation,ok :=claim.ObjectMeta.Annotations[annVolumeAlreadyProvisioned]; ok && annotation== "yes" {

A little suggestion. :=)

**Comment by jsafrane** [comment]
> Would it be acceptable solution if I find a way to to encrypt annStoredVolumeData?

It looks extremely ugly to me. You bring security somewhere it should not be. It opens whole new can on worms, e.g. you need a way how to prevent reply attacks on different PVC.

IMO, internal provisioners are quite fine as they are now. And the external ones have wide variety of ways how to fix themselves, starting from increasing the timeout or by using CRDs in their own namespace.

Note that deleting unwanted PVs is quite complex. Current code prefers deleting volumes to save space and keep the volumes in storage backend in sync with PVs. This PR may leave orphan volumes in the storage backend without PVs for them in case user deletes PVC between provisioning retries. Both ways have its pros and cons.

**Comment by sbezverk** [comment]
@jsafrane Thanks for the comments. One question though, should the external provisioner be in sync with the logic of the in-tree pv-controller? If they behave differently (with suggested changes to only external provisioner), it might result different experience for a user.  I am not sure if it was a goal to provide seamless  user experience whether they use in-tree or out of tree controllers.

**Comment by liggitt** [comment]
> @jsafrane Would it be acceptable solution if I find a way to to encrypt annStoredVolumeData?

a couple thoughts:
1. shouldn't CreateVolume be idempotent? if the controller crashes after calling CreateVolume but before creating the PV, we shouldn't end up with orphaned volumes.
2. can the controller pre-create the PV in Pending state, then do the CSI provision call, then update the PV with capacity/attribute data from the results of the CreateVolume call and update the PV state?

**Comment by sbezverk** [comment]
@liggitt It should not, (really depends on CSI driver implementation) but even for `hostpath` driver we do check if request comes for already existing volume and if it is the case  no new volume gets created and in addition  external provisioner will generate the same volume name in CreateVolume request for the same PVC.
WRT point 2. I am building PoC to see how it behaves in large scale env.

**Comment by jsafrane** [comment]
> 2. can the controller pre-create the PV in Pending state, then do the CSI provision call, then update the PV with capacity/attribute data from the results of the CreateVolume call and update the PV state?

We did that in the first release of alpha dynamic provisioning and it has proven to be unreliable and error prone. For example, this PV will get bound to a PVC and scheduler will schedule pods that use this PVC, assuming it has complete topology labels (which it does not have) and A/D controller will try to attach it or kubelet will try to mount it.

Sure, we could extend PV controller not to bind to such PVs, but it IMO breaks API.

**Comment by jsafrane** [comment]
Another problem we had: such incomplete PVs go through admission plugin. We have a plugin that fills topology labels and it got confused by PVs with no real volumes behind. We can fix that easily, however, it again shows that it's API change and users would need to change their admission handlers too.

**Comment by liggitt** [comment]
does requiring CreateVolume to be idempotent resolve the need to delete it if writing the PV encounters an error?

**Comment by jsafrane** [comment]
In-tree provisioner calls `Delete()` because there is no PV for the volume yet:

1. User creates a PVC
2. Driver creates a volume
3. Provisioner fails to save the PV
4. [Provisioner deletes the volume]
5. User deletes the PVC.

After 5., no provisioning is called, because there is nothing to provision - Kubernetes does not need the PV at this time. If Kubernetes did not delete the volume at 4., the volume would be never deleted.

**Comment by liggitt** [comment]
> In-tree provisioner calls Delete() because there is no PV for the volume yet:
> 
>  1.   User creates a PVC
>  2.   Driver creates a volume
>  3.   Provisioner fails to save the PV
>  4.   [Provisioner deletes the volume]
>  5.   User deletes the PVC.
> 
> After 5., no provisioning is called, because there is nothing to provision - Kubernetes does not need the PV at this time. If Kubernetes did not delete the volume at 4., the volume would be never deleted.

I expected something like this:
* when 3 fails, the provisioner should requeue the pvc to be processed
* the workqueue would have sufficient information to locate the pvc, the pv, and the created volume
* when syncing, if both the pv and pvc no longer exist, the provisioner would delete the created volume (tolerating a "not found" error)

**Comment by msau42** [comment]
To handle the case where both PVC and PV are missing, the provisioner would need to keep an in-memory cache of created volumes, which would be lost on restarts.

**Comment by liggitt** [comment]
> To handle the case where both PVC and PV are missing, the provisioner would need to keep an in-memory cache of created volumes, which would be lost on restarts.

Then we're back to persisting local state prior to calling CreateVolume. The PV object seems the most coherent object to do that on.

**Comment by jsafrane** [comment]
> Then we're back to persisting local state prior to calling CreateVolume. The PV object seems the most coherent object to do that on.

And we're back at API breakage. Until now, PVs were only fully provisioned volumes, ready for binding and scheduling. With PVs for not fully provisioned volumes we need to change at least PV controller (not to bind PVC until provisioning is complete), scheduler (to wait for the PV to be fully provisioned and get topology labels in case someone force-binds PVC) and kubelet (to do the same when pod is scheduled directly, e.g. by DaemonSet). There is unknown number of external components that may need this change too, and IMO this counts as API breakage.

**Comment by sbezverk** [comment]
@jsafrane imho forcing storage backend to create/delete/create/delete volumes as a result of  API server related issues is not right, each subsystem should deal with its issues  internally,  minimizing exposure to other subsystems. If changing API, not sure why it is consider breaking, makes api subsystem more robust, I think it should be explored.

**Comment by wongma7** [comment]
Yeah this is not worth breaking pv api over, it is complicated enough, adding another phase before `Available` is imo out of the question.

I believe the motivation for this pr is ultimately https://github.com/kubernetes-csi/external-provisioner/issues/68. This supposed issue of PVs failing to save and wasting storage backend create API calls is just a symptom of the true problem https://github.com/kubernetes-csi/external-provisioner/issues/68 causing the API server throttling in the first place. Otherwise we have 0 evidence that a PV failing to save 5 times in a row is a common enough occurrence that we need to change this code.

**Comment by vladimirvivien** [comment]
@wongma7 should this be closed ?

**Comment by wongma7** [comment]
@vladimirvivien yes

**Pull request closed without merging**